mopidy_soundcloud/soundcloud.py

```diff
@@ -348,14 +356,84 @@ def parse_track(self, data, remote_url=False):
+                self.public_client_id = match.group(1)
+                logger.debug(
+                    f"Updated SoundCloud public client id to: {self.public_client_id}"
+                )
+                return
+
+    def _get_public_stream(self, progr_stream):
+        params = [("client_id", self.public_client_id)]
+        return self.public_stream_client.get(progr_stream, params=params)
+
+    @staticmethod
+    def parse_fail_reason(reason):
+        return "" if reason == "Unknown" else f"({reason})"
+
     @cache()
-    def get_streamble_url(self, url):
-        req = self.http_client.head(streamble_url(url, self.CLIENT_ID))
+    def get_streamable_url(self, sharing, permalink_url, stream_url):
+
+        if self.public_client_id is None:
+            self._update_public_client_id()
+
+        progressive_urls = {}
+        if sharing == "public" and self.public_client_id is not None:
+            res = self.public_stream_client.get(permalink_url)
+
+            for html_substring in res.text.split('"'):
+                if html_substring.endswith("preview/progressive"):
+                    progressive_urls["preview"] = html_substring
+                elif html_substring.endswith("stream/progressive"):
+                    progressive_urls["stream"] = html_substring
+
+                if progressive_urls.get("preview"):
+                    if progressive_urls.get("stream"):
+                        break
+
+            if progressive_urls.get("stream"):
+                stream = self._get_public_stream(progressive_urls["stream"])
+                if stream.status_code in [401, 403, 429]:
+                    self._update_public_client_id()  # refresh public client id once
+                    stream = self._get_public_stream(progressive_urls["stream"])
+
+                try:
+                    return stream.json().get("url")
+                except Exception as e:
+                    logger.info(
+                        "Streaming of public song using public client id failed, "
+                        "trying with standard app client id.."
+                    )
+                    logger.debug(
+                        f"Caught public client id stream fail:\n{str(e)}"
+                        f"\n{self.parse_fail_reason(stream.reason)}"
+                    )
+
+        # ~quickly yields rate limit errors
+        req = self.http_client.head(f"{stream_url}?client_id={self.CLIENT_ID}")
         if req.status_code == 302:
             return req.headers.get("Location", None)
         elif req.status_code == 429:
-            reason = "" if req.reason == "Unknown" else f" ({req.reason})"
-            logger.warning(f"SoundCloud daily rate limit exceeded{reason}")
+            logger.warning(
+                "SoundCloud daily rate limit exceeded "
+                f"{self.parse_fail_reason(req.reason)}"
+            )
+            if progressive_urls.get("preview"):
```

Comment:
I'm hoping it's not going to be confusing to use the preview stream when the public and private streams don't work. Personally I can't think of a situation where I'd be interested in it.

Reply:
My reasoning was to play at least something if possible. But I can imagine people opting for not playing anything. If anyone has compelling arguments for some policy, it can always be changed down the line.